### DISCUSSION TITLE
Unexpected results combining Biquad filters.

### DISCUSSION THREAD

**Discussion by PaulDevenney** [ar1st0crat/NWaves#72] [General]
According to https://github.com/ar1st0crat/NWaves/wiki/Filters#combining-filters, cascaded filters are equivalent to chained `ApplyTo()`, but I don't observe this behaviour. I assume I'm doing something wrong and would be greatful for advice.

```
 var fl = 300.0;
            var fh = 600.0;
            var frequency = Math.Sqrt(fl * fh);
            var q = frequency / (fh - fl);

            var peakFilter = new PeakFilter(frequency / sampleRate, q, 5);
            var highpassFilter = new HighPassFilter(100.0 / sampleRate);
            var highShelfFilter = new HighShelfFilter(10000.0/sampleRate,1, 5);

            //Why are eqdSignal and cascadedSignal different?
            var eqdSignal = highShelfFilter.ApplyTo(peakFilter.ApplyTo(highpassFilter.ApplyTo(signal)));
            
            var cascadeFilter = new IirFilter(highpassFilter.Tf * peakFilter.Tf * highShelfFilter.Tf);
            var cascadedSignal = cascadeFilter.ApplyTo(signal);
```
leads to the following; 
blue = FFTd signal
orange = FFTd eqdSignal
red = FFTd cascadedSignal

![image](https://user-images.githubusercontent.com/6560938/210574343-0d49e7dc-522b-4e6b-8096-4b8edc5e84b6.png)

The orange line (`highShelfFilter.ApplyTo(peakFilter.ApplyTo(highpassFilter.ApplyTo(signal)))`) is what I would have expected to see as a result.

**Comment by ar1st0crat** [comment]
Hi, you're doing everything right essentially. The two approaches are equivalent conceptually, but technically combination of simpler filters will always work better (numerically, less round-off errors). Also, if you play around with parameters of your filters you'll notice that output discrepancies will be different depending on those parameters and sometimes they'll be pretty insignificant.
So, the advice is to use separate filters. If you want to make sure that the convolution of TFs also works OK, apply IIR filter to the equivalent signal with ```double``` precision.

```C#
using NWaves.Filters.Base64;
using NWaves.Utils;

var cascadeFilter = new IirFilter64(highpassFilter.Tf * peakFilter.Tf * highShelfFilter.Tf);
double[] cascadedSignal = cascadeFilter.ApplyTo(signal.Samples.ToDoubles());
```

PS. Also, there's a ```FilterChain``` class providing handy interface for sequences of filters. You can use it like this:

```C#
var chain = new FilterChain(new IOnlineFilter[] { highpassFilter, peakFilter, highShelfFilter });
var filteredSignal = chain.ApplyTo(signal);
```

**Reply by PaulDevenney** [reply]
Ah! Thanks so much for the positive reply. I think first time through the docs I read about `FilterChain` and second time through I thought the cascade thing was the same concept. I'd missed `IirFilter64` in there too. This is such a good SDK and I feel I'm only scratching the surface so far.

**Reply by PaulDevenney** [reply]
PS. It might be worth a minor change to the example I linked at the beginning, because it appears fairly unequivocal currently that they are just "equal" in all circumstances.
